06-classification_continued.Rmd

```diff
@@ -559,16 +559,257 @@ The overall workflow for performing K-nearest neighbour classification using `ti
+res <- res %>% mutate(base_acc = base_acc)
+plt_irrelevant_accuracies <- res %>%
+  ggplot() +
+  geom_line(mapping = aes(x=ks, y=accs, linetype="Tuned K-NN")) +
+  geom_hline(data=res, mapping=aes(yintercept=base_acc, linetype="Always Predict Benign (63%)")) +
+  labs(x = "Number of Irrelevant Predictors", y = "Model Accuracy Estimate") + 
+  scale_linetype_manual(name="Method", values = c("dashed", "solid"))
+
+plt_irrelevant_accuracies
+```
+</center>
+
+Wow! Once we add sufficiently many irrelevant predictor variables, our classifier
+does not do much better than just always guessing "benign", 
+ the most common class label in the data (63% of observations are benign).
+
+So then if we can't just throw in the kitchen sink, how do we choose which
+variables to use as predictors?  This is a very difficult question to answer in
+general, and there are a number of methods that have been developed that apply
+in particular cases of interest; see the additional resources at the end of
+this chapter to find out where you can learn more about variable selection.
+Here we will discuss *best subset selection,* which is applicable to any 
+classification method (K-NN or otherwise), and works when you have a small
+number of total variables from which you want to keep only the relevant ones (to
+avoid the problem in the example above). Best subset selection works 
+as follows. If you have $m$ predictors, you: 
+
+1. create a separate model for every possible subset of predictors,
+2. tune each one using cross validation, and 
+3. finally pick the subset of predictors that gives you the highest estimated prediction accuracy.  
+
+Note that best subset selection becomes very slow when you have even a moderate
+number of predictors to choose from. This is because the number of possible predictor subsets
+grows very quickly with the number of predictors. For example, if we have $m=2$ predictors---let's call
+them A and B---then we have 3 variable sets to try: A alone, B alone, and finally A
+and B together. If we have $m=3$ predictors---A, B, and C---then we have 7
+to try: A, B, C, AB, BC, AC, and ABC. In general, the number of models
+we have to train for $m$ predictors is $2^m-1$; in other words, when we 
+get to $m=10$ predictors we have over *one thousand* models to train, and 
+at $m=20$ predictors we have
+over *one million* models to train! If you have more than about 10 predictors,
+you should first use your scientific understanding of the data
+to choose the 10 or so predictors that you think are most meaningful, and then
+run best subset selection within those 10. Alternatively, you can use a different
+predictor variable selection method; see the additional resources at the end
+of this chapter. 
+
+We now turn to implementing best subset selection in R.
+Unfortunately there is no built-in way to do this using the `tidymodels` framework,
+so we will have to code it ourselves. First we will use the `select` function
```

Comment:
This is a lot of code... I would do this myself (not thinking of this course) in a similar way, but use nested data frames, `map`, and write a function to get the accuracy - more in the `tidyverse` style of doing things. The nested data frames and `map` are something we do use in the clustering chapter, but what we are doing here is more complicated... 

I will code up my `tidyverse` style of this and paste it below to compare. But I think both solutions are more complicated than what we want in this book, and the idea we discussed earlier of abstracting this to an R package is probably the right move.

Reply:
Here is how I would roughly write this in the tidyverse style:

```
library(tidyverse)
library(tidymodels)

get_model_accuracy <- function(model_formula_predictors) {
  # create a model specification
  knn_spec <- nearest_neighbor(weight_func = "rectangular", neighbors = tune()) %>%
    set_engine("kknn") %>%
    set_mode("classification")
  # create a 5-fold cross-validation object
  cancer_vfold <- vfold_cv(cancer_subset, v = 5, strata = Class)
  # create a recipe from the model string
  cancer_recipe <- recipe(as.formula(paste("Class ~", model_formula_predictors)), data = cancer_subset) %>%
    step_scale(all_predictors()) %>%
    step_center(all_predictors())
  # tune the K-NN classifier with these predictors, and collect the accuracy for the best K
  acc <- workflow() %>%
    add_recipe(cancer_recipe) %>%
    add_model(knn_spec) %>%
    tune_grid(resamples = cancer_vfold, grid = 10) %>%
    collect_metrics() %>%
    filter(.metric == "accuracy") %>%
    summarize(mx = max(mean))
  acc$mx %>% unlist()
}

cancer <- read_csv("data/wdbc.csv")
cancer_train <- cancer
cancer_subset <- cancer_train %>% select(Class, Smoothness, Concavity, Symmetry, Area)

possible_predictors = colnames(cancer_subset %>% select(-Class))

best_subset <- tibble(n_pred = 2:4, 
                      poss_pred = list(possible_predictors)) %>% 
  mutate(combinations = map2(poss_pred, n_pred, combn, simplify=FALSE)) %>% 
  unnest(combinations) %>% 
  mutate(formula_predictors = map(combinations, ~paste(., collapse="+"))) %>% 
  unnest(formula_predictors) %>% 
  mutate(accuracy = map_dbl(formula_predictors, get_model_accuracy))
```
I might be able to simplify it a bit more, but not much. I think it's too complicated for the book, so I still vote to abstract it to a package.

Reply:
I can simplify it to this with the new `dplyr` `rowwise` function:

```
library(tidyverse)
library(tidymodels)

get_model_accuracy <- function(model_formula_predictors) {
  # create a model specification
  knn_spec <- nearest_neighbor(weight_func = "rectangular", neighbors = tune()) %>%
    set_engine("kknn") %>%
    set_mode("classification")
  # create a 5-fold cross-validation object
  cancer_vfold <- vfold_cv(cancer_subset, v = 5, strata = Class)
  # create a recipe from the model string
  cancer_recipe <- recipe(as.formula(paste("Class ~", model_formula_predictors)), data = cancer_subset) %>%
    step_scale(all_predictors()) %>%
    step_center(all_predictors())
  # tune the K-NN classifier with these predictors, and collect the accuracy for the best K
  acc <- workflow() %>%
    add_recipe(cancer_recipe) %>%
    add_model(knn_spec) %>%
    tune_grid(resamples = cancer_vfold, grid = 10) %>%
    collect_metrics() %>%
    filter(.metric == "accuracy") %>%
    summarize(mx = max(mean))
  acc$mx %>% unlist()
}

cancer <- read_csv("data/wdbc.csv")
cancer_split <- initial_split(cancer_train, prop = 0.75, strata = Class)
cancer_subset <- training(cancer_split) %>% 
  select(Class, Smoothness, Concavity, Symmetry, Area)

possible_predictors = colnames(cancer_subset %>% select(-Class))

best_subset <- tibble(n_pred = 2:4) %>% 
  rowwise() %>% 
  mutate(combo = list(combn(possible_predictors, n_pred, simplify=FALSE))) %>% 
  unnest(combo) %>% 
  rowwise() %>% 
  mutate(formula_predictors = paste(combo, collapse="+"),
         accuracy = get_model_accuracy(formula_predictors))
```

Reply:
@ttimbers the chapter now uses forward selection -- any idea how to simplify that?

Reply:
closing this thread since we've moved to forward selection and @ttimbers is going to spin that out into a separate package.